Anoma/Simulator.lean

```diff
@@ -0,0 +1,155 @@
+import Anoma.Program
+import Anoma.Logic
+import Anoma.Transaction
+import AVM.Object
+import AVM.Scope
+import Mathlib.Control.Random
+
+namespace Anoma.Program
+
+structure RmState : Type 2 where
+  gen : StdGen
+  objects : Std.HashMap ObjectId Resource
+  committed : Std.HashSet Commitment
+  nullified : Std.HashSet Resource
+  logics : Std.HashMap LogicRef LogicFunction
+
+  logs : List String
+
+abbrev RmState.ini (logics : Std.HashMap LogicRef LogicFunction) (gen : StdGen := mkStdGen 0) : RmState :=
+  { gen
+    objects := ∅
+    nullified := ∅
+    committed := ∅
+    logics
+    logs := ∅ }
+
+/-- The evaluation monad -/
+abbrev RunM (a : Type 2) : Type 2 :=
+  EStateM (ULift Program.Error) RmState a
+
+def throw' {α : Type _} (e : Program.Error) : RunM α :=
+  throw (ULift.up e)
+
+/-- checks that consumed and created resources of the same kind are balanced -/
+def checkDelta (consumed created : List Resource) : RunM PUnit := do
+  let mkMap (l : List Resource) : Std.HashMap Resource.Kind (List Resource) := l.groupByKey (·.kind)
+  let consumedByKind := mkMap consumed
+  let createdByKind := mkMap created
+  let kinds : Std.HashSet Resource.Kind := Std.HashSet.ofList (consumedByKind.keys ++ createdByKind.keys)
+  let getQuantityOfKind (m : Std.HashMap Resource.Kind (List Resource)) (k : Resource.Kind) : Nat :=
+    m.get? k |>.getD [] |>.map (·.quantity) |>.sum
+  for kind in kinds do
+    let numConsumed := getQuantityOfKind consumedByKind kind
+    let numCreated := getQuantityOfKind createdByKind kind
+    if numConsumed == numCreated
+    then pure .unit
+    else throw' (.balanceCheck s!"numConsumed: {numConsumed}/{consumed.length}, numCreated: {numCreated}/{created.length}")
+
+def picks {A : Type u} (l : List A) : List (A × List A) :=
+  List.finRange l.length |>.map (fun i => ⟨l.get i, l.eraseIdx i⟩)
+
+def fetchLogic (ref : LogicRef) : RunM (LogicFunction) := do
+  let s <- get
+  match s.logics.get? ref with
+  | none => throw' (.missingLogic ref)
+  | some s => pure s
+
+def storeLogic (ref : LogicRef) (f : LogicFunction) : RunM PUnit := do
+  modify (fun s => {s with logics := s.logics.insert ref f})
+
+def storeCreated (created : List Resource) : RunM PUnit := do
+  let created := created.filter (·.ephemeral.not)
+  for r in created do
+    dbgTrace s!"created" (fun _ =>
+    let try val : AVM.Object.Resource.SomeValue := tryCast r.value
+    dbgTrace s!"hi: {val.uid}" (fun _ =>
+    modify (fun s => {s with objects := s.objects.insert val.uid r
+                             committed := s.committed.insert r.commitment})))
+
+def runAction (a : Action) : RunM PUnit := do
+  let units : List ComplianceUnit := a.complianceUnits
+  let witnesses : List ComplianceWitness := units.map (·.witness)
+  let consumed : List Resource := witnesses.map (·.consumedResource)
+  let created : List Resource := witnesses.map (·.createdResource)
+  let createdPicks := picks created
+  let consumedPicks := picks consumed
```

Comment:
It's actually been unclear to me whether `self` is included in the `consumed` list in RL arguments or not. I think currently most of the code assumes it is included. Probably that's why logic checks don't work. We should decide on one option and review all relevant code.

Reply:
I used this as reference:
https://specs.anoma.net/v1.0.0/arch/system/state/resource_machine/data_structures/action/index.html#to_instance

I'm not sure if the docs are still relevant though.

I've tried both options and both fail. For now I've left the one in the docs, but I can change that

Reply:
Yes, let's keep it the way it is in the docs and adapt our translation. Possibly there are some inconsistent assumptions about the self resource in different RLs we generate.